cmds/test-dev.c

```diff
@@ -0,0 +1,303 @@
+				return -1;
+			}
+		}
+	}
+	return EOK;
+}
+
+#endif
+
+
+static int cmd_testDevErase(int major, int minor, addr_t addr, size_t length)
+{
+	ssize_t res;
+
+	res = devs_erase(major, minor, addr, length, 0);
+	if (res < 0) {
+		log_error("\nErase failed %zd\n", res);
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+static int cmd_testDevCheck(int major, int minor, addr_t addr, size_t length)
+{
+	ssize_t res;
+	size_t rsz, i, chunk;
+	u8 buf[BUF_SIZE];
+
+	for (rsz = 0; rsz < length; rsz += res) {
+		chunk = min((length - rsz), sizeof(buf));
+		res = devs_read(major, minor, addr + rsz, buf, chunk, TEST_DEV_TIMEOUT_MS);
+		if (res < 0) {
+			log_error("\nCan't read data %zd\n", res);
+			return res;
+		}
+		for (i = 0; i < chunk; i++) {
+			if (buf[i] != ((rsz + i) & 0xffu)) {
+				log_error("\nDiscrepancy at index %zu, expected %zu got %u\n", rsz + i, i, buf[i]);
+				return -1;
+			}
+		}
+	}
+	return EOK;
+}
+
+
+static int cmd_doTestDev(int major, int minor, addr_t addr, size_t length, u8 erase)
+{
+	ssize_t res;
+
+	res = devs_check(major, minor);
+	if (res < 0) {
+		log_error("\nInvalid device %zd\n", res);
+		return -EINVAL;
+	}
+
+	if ((erase & ERASE_BEFORE) != 0u) {
+		res = cmd_testDevErase(major, minor, addr, length);
+		if (res < 0) {
+			return res;
+		}
+	}
+
+	res = cmd_testDevWrite(major, minor, addr, length);
+	if (res < 0) {
+		if ((erase & ERASE_AFTER) != 0u) {
+			(void)cmd_testDevErase(major, minor, addr, length);
+		}
+
+		return res;
+	}
+
+	res = devs_sync(major, minor);
+	if (res < 0) {
+		lib_printf("\nSync failed %zd\n", res);
+		if ((erase & ERASE_AFTER) != 0u) {
+			(void)cmd_testDevErase(major, minor, addr, length);
+		}
+
+		return res;
+	}
+
+	res = cmd_testDevCheck(major, minor, addr, length);
+	if (res < 0) {
+		(void)cmd_testDevRead(major, minor, addr, length);
+		if ((erase & ERASE_AFTER) != 0u) {
+			(void)cmd_testDevErase(major, minor, addr, length);
+		}
+
+		return res;
+	}
+
+#if (CMD_TESTDEV_READ_BIG_BUF_SIZE != 0)
+	res = cmd_testDevCheckReadBig(major, minor, addr, length);
+	if (res < 0) {
+		return res;
+	}
+#endif
+
+	if ((erase & ERASE_AFTER) != 0u) {
+		res = cmd_testDevErase(major, minor, addr, length);
+		if (res < 0) {
+			return res;
+		}
+	}
+
+	log_info("\ntest-dev: Success!\n");
+
+	return EOK;
+}
+
+
+static int cmd_testDev(int argc, char *argv[])
+{
+	int major, minor, opt, err;
+	u8 erase = 0;
+	char *endptr;
+	addr_t start;
+	size_t length;
+
+	for (;;) {
+		opt = lib_getopt(argc, argv, "eE");
+		if (opt < 0) {
+			break;
+		}
+		switch (opt) {
+			case 'e':
+				erase |= ERASE_BEFORE;
+				break;
+
+			case 'E':
+				erase |= ERASE_AFTER;
+				break;
+
+			default:
+				cmd_testDevUsage();
+				return CMD_EXIT_FAILURE;
+		}
+	}
+
+	if ((optind + 2) > argc) {
+		lib_consolePuts("\nInvalid arguments.\n");
+		cmd_testDevUsage();
+		return CMD_EXIT_FAILURE;
+	}
+
+	major = lib_strtoul(argv[optind], &endptr, 0);
```

Comment:
Just a suggestion - it might be nicer to use getopt all the way, i.e. to pass all parameters this way:
-d major.minor
-s start (can be zero by default?)
-l lenght (can be whole device by default?)

Reply:
Good idea, though I don't see any api to see length of the device so I think this parameter has to be mandatory.